File: src/201_300/0208_implement-trie-prefix-tree/implement-trie-prefix-tree.java
class TrieNode {
	TrieNode[] childrens = new TrieNode[26];
	char val;
	boolean is_end = false;
}

class Trie {
	final TrieNode root;
    /** Initialize your data structure here. */
    public Trie() {
        root = new TrieNode();
        root.val = ' ';
    }
    
    /** Inserts a word into the trie. */
    public void insert(String word) {
        TrieNode node = root;
        for(int i=0;i<word.length();i++) {
        	char c = word.charAt(i);
        	if( node.childrens[c-'a']==null ) {
        		node.childrens[c-'a'] = new TrieNode();
        	}
        	node = node.childrens[c-'a'];
        }
        node.is_end = true;
    }
    
    /** Returns if the word is in the trie. */
    public boolean search(String word) {
        TrieNode node = root;
        for(int i=0;i<word.length();i++) {
        	char c = word.charAt(i);
        	if( node.childrens[c-'a']== null ) {
        		return false;
        	}
        	node = node.childrens[c-'a'];
        }
        return node.is_end;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(String prefix) {
    	TrieNode node = root;
    	for(int i=0;i<prefix.length();i++) {
    		char c = prefix.charAt(i);
    		if( node.childrens[c-'a']==null ) {
    			return false;
    		}
    		node = node.childrens[c-'a'];
    	}
    	return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */

File: src/201_300/0208_implement-trie-prefix-tree/implement-trie-prefix-tree.py
class Trie(object):
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = {}
        self.is_end = "#"

    def insert(self, word):
        """
        Inserts a word into the trie.
        :type word: str
        :rtype: None
        """
        node = self.root
        for c in word:
            node = node.setdefault(c,{})
        node[self.is_end] = self.is_end

    def search(self, word):
        """
        Returns if the word is in the trie.
        :type word: str
        :rtype: bool
        """
        node = self.root
        for c in word:
            if(not node.has_key(c)):
                return False
            node = node[c]
        return node.has_key(self.is_end) 

    def startsWith(self, prefix):
        """
        Returns if there is any word in the trie that starts with the given prefix.
        :type prefix: str
        :rtype: bool
        """
        node = self.root
        for c in prefix:
            if(not node.has_key(c)):
                return False
            node = node[c]
        return True
                

# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)  